src/velocityper.1
.\" #
.\" # Copyright (c) 2023, Phil Shafer
.\" # All rights reserved.
.\" # This SOFTWARE is licensed under the LICENSE provided in the
.\" # LICENSE file. By downloading, installing, copying, or
.\" # using the SOFTWARE, you agree to be bound by the terms of that
.\" # LICENSE.
.\" # Phil Shafer, May 2023
.\"
.Dd May 29, 2023
.Dt XO 1
.Os
.Sh NAME
.Nm velocityper
.Nd stuff data into a terminal input buffer
.Sh SYNOPSIS
.Nm
.Op Fl options
.Op Ar argument...
.Sh DESCRIPTION
.Pp
The
.Nm
utility allows the user to place character data into a terminal's
input buffer, as if they were entered from the keyboard, making
.Nm
ideal for recording demonstration videos.
Commands and text input can be essentially pre-recorded, avoiding
typing mistakes and incorrect data, improving the quality and
efficiency of the videos.
There are a variety of timers suitable for tuning the pacing of the
input to make it appear natural.
.Pp
Data and options can be placed in input files to facilitate lengthy or
complex recordings, and pauses can be incorporated into the session,
with the user being prompted on a different terminal to proceed.
.Pp
.Nm
uses a normal terminal control operation (ioctl
.Nm TIOCSTI )
to place data into the keyboard, which means the user can continue to
operate the keyboard and provide input during these pauses, allowing a
mix of "canned" and interactive content.
.Pp
.Sh UNITS
.Pp
Options that accept time values default to using milliseconds as the units,
but support the following units as suffixes:
.Bl -column "Short  " "Long Form123" "Nanoseconds"
.It "Short  " "Long Form   " " Nanoseconds"
.It "ns     " "nanoseconds " "             1"
.It "us     " "microseconds" "         1,000"
.It "ms     " "milliseconds" "     1,000,000"
.It "s      " "seconds     " " 1,000,000,000"
.El
Leading strings can be used for brevity;
for example "nano" can be used to nanoseconds. 
.Sh OPTIONS
.Pp
.Nm
supports a variety of command-line option for tuning its behavior, and
these options can appear in input files.
Several of them are better suited for such files, but are also
supported on the command line for consistency.  See the "FILE FORMAT"
section below for details.
.Pp
The following options are supported:
.Bl -tag -width indent
.\"
.\"
.It Ic -b | Ic --bump
Introduce a random addition to the intra-charater wait value given
with the
.Nm --wait
option, giving the text a means of appearing more
natural and less automated.
For example, using "--wait 40 --bump 60" means a delay of 40ms plus a
bump of 0-60ms, for  complete range of 40-100ms.
.\"
.\"
.It Ic -C <terminal> | Ic --confirm <terminal>
Provides a managing terminal used for prompting for
confirmations.
.\"
.\"
.It Ic -D | --debug
Generate internal debug information.
.\"
.\"
.It Ic -e <delay> | Ic --end <delay>
Introduce a delay before each line of data.
This can be used to allow the viewer to see a full command string
before the command is executed.
.\"
.\"
.It Ic -F | Ic --force
Ignore any confirmations requested by the
.Nm --pause-confirm
option.
.\"
.\"
.It Ic -f <filename> | Ic --file <filename>
Use the given file as content.  See the "FILE FORMAT" section below.
.\"
.\"
.It Ic -l <delay> | --line <delay>
Introduce a delay after each line of data.
.\"
.\"
.It Ic -n | Ic --dry-run
Echo the content on the current tty, but do not place it into any
terminal input buffer; delays operate normally.
This can be used to practice the demo or to evaluate timing.
It effectively turns
.Nm
into
.Xr echo 1 .
.It Ic -P <message> | Ic --pause-confirm <message>
Instructs
.Nm
to emit a message on the managing terminal and prompt for a return
before continuing, allowing the demonstration to
pause and continue as needed.
Alternatively, use
.Nm -+ message
lines in files to prompt for confirmation.
.\"
.\"
.It Ic -p <delay> | --pause <delay>
Pause the output for the period provided.
.\"
.\"
.It Ic -S <period> | --sleep <period>
Sleep immediately for the specified period.
.\"
.\"
.It Ic -s | --skip
Don't make output or do delays.  Multiple invocations of this option
toggle between skipping and not skipping.
This option can be used in input files to turn
on and off operations, effectively commenting out section of data.
.\"
.\"
.It Ic -T <capname> | --tput <capname>
Emit the terminal capability associated with the given name.  This
option is meant to mimic the
.Xr tput 1
command, but currently only implements the
.Nm clear
and
.Nm home
capabilities.
.\"
.\"
.It Ic -t <terminal> | --tty <terminal>
Provides a terminal to use for pushing input data, allowing
one terminal to handling the prompting while running
the demonstration in a second terminal.
This option requires super-user permissions,
since it opens security concerns. 
.\"
.\"
.It Ic -w <delay> | Ic --wait <delay>
Introduce a basic delay after each character, slowing the rate as
required to make the text entry approximate human typing.
Use the 
.Nm --bump
to add an random delay in addition to the value given using the
.Nm --wait
option, allowing the characters to appear in a more natural tempo.
.\"
.\"
.El
.Sh ESCAPE SEQUENCES
.Pp
The
.Nm
utility accepts format strings that can contain escape sequences,
which must be preceeded by a backslash.
The following table lists the supporting escape sequences:
.Bd -literal -offset indent
EscSeq  Operation
\\a       Emit alarm ('^G')
\\b       Emit backspace ('^H')
\\d       Emit EOF ('^D')
\\e       Emit escape ('^[')
\\f       Emit formfeed ('^L')
\\n       Emit newline ('^N')
\\p       Pause for period from --pause option
\\P       Pause for confirmation
\\r       Emit carriage return ('^R')
\\t       Emit TAB ('^')
\\u{num}  Emit UTF-8 encoding of the code point
\\xYZ     Emit hex character (0xYZ)
\\x{hex}  Emit a series of two-digit hex values
\\^N      Emit the corresponding control character (e.g. "\^G" or "\^g")
\\^[      Emit escape ('^[')
\\^\\      Emit file separator (0x1c)
\\^]      Emit group separator (0x1d)
\\^^      Emit record separator (0x1e)
\\^_      Emit unit separator (0x1f)
\\^?      Emit delete (0x7f)
\\^@      Emit NUL character (0x00)
.Ed
.Sh FILE FORMAT
.Pp
The
.Nm --file
option allows content to be placed in a file, using the following syntax:
.Bl -bullet
.\"
.It
Lines that start with a pound character ('#') are comments.
.Bd -literal -offset
# This is a comment
.Ed
.\"
.It
Lines that start with dash followed by a plus sign ("-+") are
prompt confirmations, similar to the
.Nm -P
option.
.Bd -literal -offset
# This line will prompt before running a command:
-+ before running command
.Ed
.\"
.It
Lines that start with a dash ("-") are parsed as command line options,
allowing the file to set or change option values.
.Bd -literal -offset
# These lines changes option values:
-w 10ms -e 1s --bump 100ms
--line 400ms
.Ed
.\"
.It
Lines that start with a backslash character ("\\") escape any
interpretation of that line, treating it as just normal text.
.Bd -literal -offset
\\# this is not seen as a comment
\\-- this is not seen as an option
.Ed
.\"
.It
Any other lines are text that
will be placed into the input buffer of appropriate tty.
.Bd -literal -offset
ls -l ~/bin
df ${HOME}
.Ed
.\"
.It
Newlines ("\\n") are turned into carriage returns ("\\r"),
suitable for terminal input.
.It
Trailing backslashes will escape a trailing newline, and
it will be ignored.
.Bd -literal -offset
# The first line below will not execute the command.  We'll
# sleep for 5 seconds and then enter it (using the blank line):
ls -l /usr/src\\
--sleep 5s

# Above line left intentionally blank.
.Ed
.It
The escape sequences listed above will be processed.
.Bd -literal -offset
# Turn off stuffing and waiting, set pause to one second
# and emit a series of alternating pluses and minuses:
--dry-run --wait=0 --pause=1s
+\\p\\b-\\p\\b+\\p\\b-\\p\\b+\\p\\b-\\p\\b+\\p\\b-
\\p\\b+\\p\\b-\\p\\b+\\p\\b-\\p\\b+\\p\\b-\\p\\b+\\b\\
Done
--dry-run
.Ed
.El
.Sh EXAMPLES
.Pp
This example places six lines of data into the input buffer:
.Bd -literal -offset indent
velocityper -b 70ms -w 30ms -l 150ms "one\\rtwo\\rthree\\rfour\\rfive\\rsix\\r"
.Ed
.Pp
This example places three lines of data into the input buffer, pausing
one second after each line:
.Bd -literal -offset indent
velocityper -w 5 -b 95 -p 1s "echo 1\\r\\pecho 2\\r\\pecho 3\\r"
.Ed
.Pp
This example does not stuff data ('-n') but emits on stdout
the characters "cdef" on one line and the UTF-8 smiley face
on a second line:
.Bd -literal -offset indent
velocityper -n 'one: \\x{63646566}\\ntwo: \\u{263A}\\n'
.Ed
.Pp
This example uses a file to perform a demo in another terminal:
.Bd -literal -offset indent
sudo velocityper --tty /dev/pts/1 --file my-demo.vt
.Ed
.Pp
The file would contain the full demo content:
.Bd -literal -offset indent
#
# Demo for JUNOS
#  These lines will drive a demo of the JUNOS CLI
# First we adjust the timers to human-ish values
-w 20ms -b 70ms -l 200ms
# Then we pause to wait until I'm ready to start
# Note: since lines in the file that start with "-" are options,
# these option values must be quoted as if they were on the
# command like. '-P ready to start' is not acceptable, since the
# lack of quotes makes four tokens here, not two.
# Use "-+ ready to start" to avoid quoting issues.
-P "ready to start demo"
configure private
edit protocols bgp group foo neighbor 1.2.3.4
# Pause again to explain what's about to happen
-+ bgp complete
set apply-macro foo one 1
set apply-macro foo two 2
show
# pause to allow more discussion
-+ apply complete
set apply-lock user phil
up 1
protect neighbor 1.2.3.4
# These skip lines are used to comment out a section of data,
# which can be done using comments, but if the section is
# lengthy, then adding two "--skip" lines might be easier.
--skip
show
--skip
# At each of these pauses, I can talk as well as type on
# the terminal, mixing canned and interactive content.
# But I might need to restore some state (e.g. location) before
# hitting RETURN in the other terminal.
-+ protect complete
show | compare
-+ done
.Ed
.Pp
Another example file:
.Bd -literal -offset indent
# Use this file with the -C or -t options to use a distinct
# tty for confirmations.
# This line gives some timer option values
-w 15 -b 40 -e 300
# This line issues a simple command on the tty
show system information
# Prompt to make a delay so the presenter can talk
-+ before 'show version'
# This time we pause before the newline so the presenter
# can talk about the command before the output appears.
# This requires the line to end with an escape so the
# end-of-line newline is ignored.  After the prompt, we
have a blank line, and that newline enters this command
show version\\
-+ before newline

-+ after newline
.Ed
.Sh HISTORICAL NOTES
.Pp
This command will not work under OpenBSD, due to removal of TIOCSTI:
.Bd --literal --offset indent
https://undeadly.org/cgi?action=article&sid=20170701132619
.Ed
.Pp
On many operating systems, the terminal input buffer has a
finite size limit, some as low as 1024 (e.g. MAX_INPUT on MacOS).
This may impact the ability of
.Nm
to place data into the input buffer without another process actively
reading from the terminal.
.Sh AUTHORS
.Nm
was written by
.An Phil Shafer Aq Mt phil@freebsd.org .


src/velocityper.c
/*
 * Copyright (c) 2023, Phil Shafer
 * All rights reserved.
 * This SOFTWARE is licensed under the LICENSE provided in the
 * LICENSE file. By downloading, installing, copying, or otherwise
 * using the SOFTWARE, you agree to be bound by the terms of that
 * LICENSE.
 * Phil Shafer, May 2023
 */

#include <unistd.h>
#include <stdio.h>
#include <termios.h>
#include <stdlib.h>
#include <getopt.h>
#include <string.h>
#include <fcntl.h>
#include <err.h>
#include <errno.h>
#include <time.h>
#include <ctype.h>
#include <limits.h>
#include <sys/ioctl.h>

#ifndef TRUE			/* #include <stdtrue>? */
#define TRUE 1
#endif
#ifndef FALSE
#define FALSE 0
#endif

#ifdef NOCONFIG
#define VELOCITYPER_VERSION "[unknown]" /* Fake a value */
#else
#include "config.h"
#endif /* NOCONFIG */

/*
 * We carry time internally as nanoseconds.
 */
#define NSECS_PER_uSEC 1000	/* Nanoseconds per microsecond */
#define uSECS_PER_MSEC 1000	/* Microseconds per millisecond */
#define MSECS_PER_SEC  1000	/* Milliseconds per second */

#define NSECS_PER_MSEC (NSECS_PER_uSEC * uSECS_PER_MSEC)
#define NSECS_PER_SEC (NSECS_PER_MSEC * MSECS_PER_SEC)

typedef unsigned long long nanoseconds_t;

static struct opts {
    int o_fd;			/* Holds our file descriptor */
    FILE *o_confirm;		/* Terminal to read confirmations from */
    nanoseconds_t o_bump;	/* Random 'bump' to pause */
    int o_debug;		/* Debug flag */
    nanoseconds_t o_end;	/* Time to pause before each line ends */
    char *o_file;		/* '--file' option */
    int o_force;		/* Force: skip confirmations  */
    int o_dryrun;		/* Dry run: don't confirm anything */
    nanoseconds_t o_line;	/* Time to pause after each line */
    nanoseconds_t o_pause;	/* Time to pause for '\p' */
    int o_skip;			/* Don't process anything */
    char *o_tty;		/* Terminal to prompt on */
    nanoseconds_t o_wait;	/* Time to pause after each character */
} opts;

/*
 * These next two table _MUST_ be kept in the same order.
 */
static struct option long_opts[] = {
    { "bump", required_argument, NULL, 'b' },
    { "confirm", required_argument, NULL, 'C' },
    { "debug", no_argument, NULL, 'D' },
    { "end", required_argument, NULL, 'e' },
    { "force", no_argument, NULL, 'F' },
    { "file", required_argument, NULL, 'f' },
    { "help", no_argument, NULL, 'h' },
    { "line", required_argument, NULL, 'l' },
    { "dry-run", no_argument, NULL, 'n' },
    { "parse-confirm", required_argument, NULL, 'P' },
    { "pause", required_argument, NULL, 'p' },
    { "sleep", required_argument, NULL, 'S' },
    { "skip", no_argument, NULL, 's' },
    { "tput", required_argument, NULL, 'T' },
    { "tty", required_argument, NULL, 't' },
    { "version", no_argument, NULL, 'v' },
    { "wait", required_argument, NULL, 'w' },
    { NULL, 0, NULL, 0 }
};

static const char *help_opts[] = {
    /* b */ " <delay>", "Bump the 'wait' timer by a random amount",
    /* C */ " <tty>", "Provide a terminal name for prompting confirmations",
    /* D */ "", "Enable debug output",
    /* e */ " <end>", "Provide a delay before newlines",
    /* F */ "", "Ignore confirmation requests",
    /* f */ " <file>", "Provide a file for content data",
    /* h */ "", "Display this help message",
    /* l */ " <delay>", "Set delay between lines of data",
    /* n */ "", "Do not place data in input buffer; just echo it",
    /* P */ " <msg>", "Emit a message and wait for confirmation",
    /* p */ " <delay>", "Set the 'pause' delay",
    /* S */ " <time>", "Sleep immediately for the given period",
    /* s */ "", "Do not make output or perform delays",
    /* T */ " <capname>", "Emit terminal capability ('clear' or 'home')",
    /* t */ " <tty>", "Provide a terminal for pushing input data",
    /* v */ "", "Emit version information (and exit)",
    /* w */ " <delay>", "Set the 'wait' delay between characters",
    NULL
};

static const char *program_name;

static void
print_help (void)
{
    int has_help = 1;
    char buf[128];
    struct option *opt;
    const char **help = help_opts;

    printf("Usage: %s [options...] [strings...]\n", program_name ?: "velocityper");

    for (opt = long_opts; opt->name; opt++, help += 2) {
	if (has_help && *help == NULL)
	    has_help = 0;	/* Done run out of data: SNO */

	const char *arg_info = !has_help ? "" : *help;

	snprintf(buf, sizeof(buf), "-%c%s | --%s%s",
		 opt->val, arg_info, opt->name, arg_info);

	printf("    %-30s  %s\n", buf, has_help ? help[1] : "");
    }

    exit(1);
}


/* Forward declaration */
static void handle_file (const char *fname);

static int
prefix_match (const char *str, const char *base)
{
    for ( ; *str && *base; str++, base++) {
	if (*str != *base)
	    return 0;
    }

    return (*str == '\0');
}

static nanoseconds_t
strtons (char *cp)
{
    nanoseconds_t t = 0;
    char *ep = NULL;

    t = strtoull(cp, &ep, 10);
    if ((t == 0 && errno == EINVAL) || (t == ULLONG_MAX && errno == ERANGE))
	err(1, "invalid time value: '%s'", cp);

    /* Look for a suffix */
    if (ep && *ep) {
	if (prefix_match(ep, "seconds")) {
	    t *= NSECS_PER_SEC;
	} else if (prefix_match(ep, "milliseconds") || prefix_match(ep, "ms")) {
	    t *= NSECS_PER_MSEC;
	} else if (prefix_match(ep, "microseconds") || prefix_match(ep, "us")) {
	    t *= NSECS_PER_uSEC;
	} else if (prefix_match(ep, "nanoseconds") || prefix_match(ep, "ns")) {
	    /* nothing */
	} else {
	    errx(1, "unknown time unit: '%s'", ep);
	}
    } else {
	t *= NSECS_PER_MSEC;
    }

    if (opts.o_debug)
	fprintf(stderr, "[strtoms '%s' -> %lld]\n", cp, t);

    return t;
}

static void
print_version (void)
{
    printf("%s\n", VELOCITYPER_VERSION);
    exit(0);
}

static void
do_confirm (const char *msg)
{
    char buf[BUFSIZ];
    FILE *inp = opts.o_confirm ?: stdin;
    FILE *outp = opts.o_confirm ?: stdout;

    if (msg == NULL || *msg == '\0')
	msg = "paused";

    fprintf(outp, "[%s; press enter to continue]\n>>> ", msg);
    fflush(outp);

    (void) fgets(buf, sizeof(buf), inp); /* We don't care about content */
}

static void
do_pause (nanoseconds_t p)
{
    struct timespec tv;

    tv.tv_sec = p / NSECS_PER_SEC;
    tv.tv_nsec = p % NSECS_PER_SEC;

    if (opts.o_debug)
	fprintf(stderr, "[pause (%ld.%09ld)]", tv.tv_sec, tv.tv_nsec);

    nanosleep(&tv, &tv);
}

static void
handle_char (char c)
{
    if (opts.o_debug)
	fprintf(stderr, "[char %#02x]", c);

    /* Pause before newline */
    if ((c == '\r' || c == '\n') && opts.o_end) {
	do_pause(opts.o_end);
    }

    if (opts.o_dryrun) {
	write(opts.o_fd, &c, 1); /* Write the byte to the fd (not stuffed) */

    } else if (opts.o_skip) {
	return;

    } else if (ioctl(opts.o_fd, TIOCSTI, &c) < 0) {
        perror("ioctl");
        exit(1);
    }

    if (opts.o_wait) {
        int bump = opts.o_bump ? rand() % opts.o_bump : 0;

        do_pause(opts.o_wait + bump);
    }

    /* Pause after newline */
    if ((c == '\r' || c == '\n') && opts.o_line) {
	do_pause(opts.o_line);
    }
}

/**
 * Return the mask for len bits in the first byte of a UTF-8 character
 * of length 'len'.  These bits are set to 1.
 */
static inline int
utf8_len_bits (int len)
{
    const uint64_t len_bits
        = (0xf0 << 4)
        | (0xe0 << 3)
        | (0xc0 << 2)
        | (0x00 << 1);

    return (len_bits >> len) & 0xff;
}

static inline int
utf8_to_len (wchar_t wc)
{
    if ((wc & 0x7f) == wc) /* Simple case */
        return 1;
    if ((wc & 0x7ff) == wc)
        return 2;
    if ((wc & 0xffff) == wc)
        return 3;
    if ((wc & 0x1fffff) == wc)
        return 4;
    return -1;          /* Invalid input wchar */
}

static void
handle_char_utf8 (wchar_t wc)
{
    if (opts.o_debug)
	fprintf(stderr, "[wide %#06lx]", (unsigned long) wc);

    int len = utf8_to_len(wc);

    if (len <= 1) { /* Simple case (and error case) */
        handle_char((char)(wc & 0x7f));
        return;
    }

    char len_bits = utf8_len_bits(len); /* Drop in new length bits */

    switch (len) {
    case 4:
        handle_char(len_bits | (char)(0x80 | ((wc >> 18) & 0x03)));
	len_bits = 0;
        /* fallthru */
    case 3:
        handle_char(len_bits | (char)(0x80 | ((wc >> 12) & 0x3f)));
	len_bits = 0;
        /* fallthru */
    case 2:
        handle_char(len_bits | (char)(0x80 | ((wc >> 6) & 0x3f)));
        handle_char((char)(0x80 | (wc & 0x3f)));
    }
}

static void
handle_string (char *str)
{
    char c;
    long val;

    for ( ; (c = *str); str++) {
        if (c == '\\') {
            c = *++str;
            switch (c) {
            case 0:
                return;

            case 'a':
                handle_char(0x07); /* ^G */
                break;

            case 'b':
                handle_char('\b'); /* ^H */
                break;

            case 'd':
                handle_char(0x04); /* ^D */
                break;

            case 'e':
                handle_char(0x1b); /* ESC */
                break;

            case 'f':
                handle_char(0x0c); /* ^L */
                break;

            case 'n':
                handle_char('\n');
                break;

            case 'P':
		if (!opts.o_force)
		    do_confirm(NULL);
                break;

            case 'p':
                do_pause(opts.o_pause);
                break;

            case 'r':
                handle_char('\r');
                break;

            case 't':
                handle_char('\t');
                break;

	    case 'u':
		if (str[1] == '{') {
		    wchar_t wc = strtol(str + 2, &str, 0x10);
		    if ((wc != 0 || errno != EINVAL) && *str == '}') {
			handle_char_utf8(wc);
		    }
		}
		break;

	    case 'x':
		if (str[1] == '{') {
		    for (str += 2; *str; str += 2) {
			if (!isxdigit(str[0]) || !isxdigit(str[1]))
			    break;
			
			char num[3] = { str[0], str[1], 0 };
			val = strtol(num, NULL, 0x10);
			if (val != 0 || errno != EINVAL)
			    handle_char(val);
		    }

		} else {
		    if (isxdigit(str[1]) && isxdigit(str[2])) {
			char num[3] = { str[1], str[2], 0 };
			val = strtol(num, NULL, 0x10);
			if (val != 0 || errno != EINVAL)
			    handle_char(val);
			str += 2;
		    }
		}
		break;

	    case '^':
		str += 1;
		val = *str - ('A' - 1); /* "\^G" */
		if (!(0 < val && val < 0x1b)) {
		    val = *str - ('a' - 1); /* "\^g" */
		    if (!(0 < val && val < 0x1b)) {
			switch(*str) {
			case '[': val = 0x1b; break; /* Escape */
			case '\\': val = 0x1c; break; /* File separator */
			case ']': val = 0x1d; break; /* Group separator */
			case '^': val = 0x1e; break; /* Record separator */
			case '_': val = 0x1f; break; /* Unit separator */
			case '?': val = 0x7f; break; /* Delete */
			case '@': val = 0; break;	 /* NUL */
			default: val = ' ';		 /* Nothing better? */
			}
		    }
		}

		handle_char(val);
		break;

            default:
                handle_char(c);
            }
        } else {
            handle_char(c);
        }
    }
}

/*
 * We don't really want to require ncurses and don't want the heartache
 * of initscr() and locale and all that pain.  So we stuff our fingers in
 * our ears and rest in the "knowledge" that vt100 won the terminal wars,
 * and make vt100-compatible output.  What'd you say?  I can't hear you...
 */
static void
handle_tput (const char *optarg)
{
    static const char clear[] = {
	0x1b, 0x5b, 0x48, 0x1b, 0x5b, 0x32, 0x4a, 0x1b, 0x5b, 0x33, 0x4a, 0
    };
    static const char home[] = { 0x1b, 0x5b, 0x48, 0 };
    const char *str = NULL;

    if (strcmp(optarg, "clear") == 0)
	str = clear;
    else if (strcmp(optarg, "home") == 0)
	str = home;
    else
	return;
	
    if (opts.o_dryrun) {
	char c;
	for ( ; (c = *str); str++)
	    handle_char(c);
    } else {
	/*
	 * These characters are written directly to the terminal, not
	 * stuffed into the terminal input stream.
	 */
	if (opts.o_debug)
	    fprintf(stderr, "[tput %s (%lu)]",
		    optarg, (unsigned long) strlen(str));

	(void) write(opts.o_fd, str, strlen(str));
    }
}

static void
process_argv (int ac, char **av)
{
    int rc;
    optind = 1;			/* Reset to allow restarting */
    char *msg = NULL;
    nanoseconds_t val;

    while ((rc = getopt_long(ac, av, "b:C:De:Ff:hl:nP:p:S:sT:t:vw:",
                                long_opts, NULL)) != -1) {
        switch (rc) {
        case 'b':
	    opts.o_bump = strtons(optarg);
            break;

	case 'C':
	    if (opts.o_confirm)
		fclose(opts.o_confirm);

	    opts.o_confirm = fopen(optarg, "r+");
	    if (opts.o_confirm == NULL)
		err(1, "could not open confirmation: '%s'", optarg);
	    break;

        case 'D':
            opts.o_debug = !opts.o_debug;
            break;

        case 'e':
	    opts.o_end = strtons(optarg);
            break;

        case 'F':
            opts.o_force = !opts.o_force;
            break;

        case 'f':
	    if (opts.o_file)
		err(1, "file name is already provided: '%s'", opts.o_file);
	    opts.o_file = strdup(optarg);
            break;

        case 'h':
	    print_help();
	    break;

        case 'l':
	    opts.o_line = strtons(optarg);
            break;

        case 'n':
            opts.o_dryrun = !opts.o_dryrun;
            break;

	case 'P':
	    msg = strdup(optarg);
	    break;

        case 'p':
	    opts.o_pause = strtons(optarg);
            break;

	case 'S':
	    val = strtons(optarg);
	    if (val)
		do_pause(val);
	    break;

        case 's':
            opts.o_skip = !opts.o_skip;
            break;

	case 'T':
	    handle_tput(optarg);
	    break;

        case 't':
            opts.o_tty = strdup(optarg);
            opts.o_fd = open(optarg, O_WRONLY);
            if (opts.o_fd < 0)
                err(1, "cannot open terminal: '%s'", optarg);
	    break;

	case 'v':
	    print_version();
	    break;

        case 'w':
	    opts.o_wait = strtons(optarg);
            break;

        default:
            print_help();
        }
    }

    if (msg) {
	do_confirm(msg);
	free(msg);
    }
}

static char *
strskipws (char *cp)
{
    for ( ; *cp; cp++) {
	if (!isspace(*cp))
	    return cp;
    }

    return NULL;
}

static char *
strskiptows (char *cp)
{
    char in_quote = 0;
    char *out = NULL;
    int skip = FALSE;

    for ( ; *cp; cp++) {
	if (*cp == '\\') {
	    if (cp[1] == '\0')
		return cp;

	    if (out == NULL)
		out = cp; /* Start overwriting */

	    cp += 1;

	} else if (*cp == in_quote) {
	    in_quote = 0;
	    skip = TRUE;	/* Skip this trailing quote */

	} else if (*cp == '"' || *cp == '\'') {
	    in_quote = *cp;

	    if (out == NULL)
		out = cp; /* Start overwriting */

	    skip = TRUE;	/* Skip this leading quote */

	} else if (in_quote) {
	    /* nothing */

	} else if (isspace(*cp)) {
	    if (out && out != cp)
		*out = '\0';	/* Terminate rewritten string */

	    return cp;
	}

	if (out && !skip)
	    *out++ = *cp;

	skip = FALSE;		/* Reset skip flag */
    }

    if (out)
	*out = '\0';

    return NULL;
}

static int
build_argv (char **av, int avsize, char *cp)
{
    static char dummy[] = "av0";
    int ac = 0;

    av[ac++] = dummy;

    while (ac < avsize - 1) {
	cp = strskipws(cp);
	if (cp == NULL)
	    break;

	av[ac++] = cp;
	cp = strskiptows(cp);
	if (cp == NULL)
	    break;

	*cp++ = '\0';
    }

    av[ac] = NULL;
    return ac;
}

/**
 * Process a file for the kdb input buffer.  The syntax is:
 *
 *   # lines that start with '#' are comments
 *   # blank lines are ignored
 *   # lines that start with '-' are command line options:
 *   -p 1000 --wait=100
 *   # lines that start with '\' escape these:
 *   \# the pound sign here is stuffed
 *   \\ and this backslash is as well
 *   # any other lines are stuffed
 *   # the trailing newline is turned into a RETURN ('\r')
 *   # to avoid this, end the line with a backslash
 *   # other backslash characters are support:
 *   #    '\b', '\n', '\p', '\r', '\t', etc
 *   this turns into \
 *   one line\b\b
 */
static void
handle_file (const char *fname)
{
    const int max_av = 32;
    char *av[max_av];
    int ac;
    char buf[BUFSIZ], *cp, *ep;
    FILE *fp = fopen(fname, "r");

    if (fp == NULL)
        err(1, "could not open file '%s'", fname);

    for (;;) {
        if (fgets(buf, sizeof(buf), fp) == NULL)
            break;

        ep = buf + strlen(buf);
        cp = buf;

        if (*cp == '#') {
	    /* If the very first character is '#', it's a comment */
	    continue;
	} else if (cp[0] == '-' && cp[1] == '+') {
	    /* The full line is a confirmation prompt */
	    
	    if (ep[-1] == '\n')	/* Trim trailing newline */
		ep[-1] = '\0';

	    cp = strskipws(cp + 2);
	    do_confirm(cp);

	} else if (*cp == '-') {
	    /* options */
	    if (ep != buf && ep[-1] == '\n')
		ep[-1] = '\0';

	    ac = build_argv(av, max_av, cp);
	    if (ac <= 0)
		continue;

	    process_argv(ac, av);

	    if (av[optind] != NULL) {
		fprintf(stderr, "[warning: extra argument '%s' (ignored)]\n",
			av[optind]);
	    }

	} else {
	    if (ep != buf && ep[-1] == '\n') {
		if (ep > buf + 1 && ep[-2] == '\\')
		    ep[-2] = '\0';
		else ep[-1] = '\r';
	    }

	    /*
	     * If the first character is '\', we skip it; this allows us to
	     * have "\# not a comment" lines.
	     */
            if (cp[0] == '\\')
                cp += 1;
            handle_string(cp);
            continue;
        }
       
    }
}

int
main (int argc, char **argv)
{
    int i;

    program_name = argv[0];
    opts.o_fd = 1;

    srand(time(NULL));

    process_argv(argc, argv);

    argc -= optind;
    argv += optind;

    for (i = 0; i < argc; i++) {
        if (i > 0)
             handle_char(' ');
        handle_string(argv[i]);
    }

    while (opts.o_file) {
	handle_file(opts.o_file);

	free(opts.o_file);
	opts.o_file = NULL;
    }

    exit(0);
}
